utils/retrieve-state/src/lib.rs

```diff
@@ -289,50 +337,369 @@ pub fn put_block_with_deploys(
+                }
+                Ok(PeerMsg::NoBytes { trie_key, peer }) => {
+                    println!(
+                        "got no bytes for key {} from peer {} (will query for it again...)",
+                        trie_key, peer
+                    );
+                    outstanding_trie_keys.push(trie_key);
+                }
+                Err(PeerError::BadData {
+                    trie_key,
+                    peer_address,
+                    error,
+                }) => {
+                    println!(
+                        "got bad data at key {} from peer from peer {} error {:?}",
+                        trie_key, peer_address, error
+                    );
+                    outstanding_trie_keys.push(trie_key);
+                }
+                Err(PeerError::TimedOut {
+                    trie_key,
+                    peer_address,
+                    error,
+                }) => {
+                    println!(
+                        "timed out asking for trie at key {} from peer {} error {:?}",
+                        trie_key, peer_address, error
+                    );
+                    outstanding_trie_keys.push(trie_key);
+                }
+                Err(PeerError::FailedRpc {
+                    trie_key,
+                    peer_address,
+                    error,
+                }) => {
+                    println!(
+                        "rpc failed to get key {} from peer {} error {:?}",
+                        trie_key, peer_address, error
+                    );
+                    outstanding_trie_keys.push(trie_key);
+                }
+                Err(PeerError::AddrParseError(addr_parse_error)) => {
+                    panic!("failed to parse address {:?}", addr_parse_error);
+                }
+                Ok(PeerMsg::GetTrie(_)) => panic!("GetTrie should not be sent by workers"),
+            }
+        }
+
+        // Finishing state, we downloaded all the intermediate tries we found, as well as the state
+        // root we started with.
+        if tries_downloaded == total_tries_count + 1 {
+            let missing =
+                engine_state.missing_trie_keys(CorrelationId::new(), vec![state_root_hash])?;
+
+            if missing.is_empty() {
+                println!("state root has no missing descendants.");
+                println!(
+                    "tries downloaded: {downloaded}/{total}",
+                    downloaded = tries_downloaded,
+                    total = total_tries_count,
+                );
+                break;
+            } else {
+                println!(
+                    "state root still has {} missing descendants. {:#?}",
+                    missing.len(),
+                    in_flight_counters,
+                );
+            }
+        }
+    }
+
+    for (_address, peer) in peer_map.drain() {
+        peer.shutdown.send(()).unwrap();
+        peer.join_handle.await.unwrap();
```

Comment:
Would it be a bit faster to first send to _all_ of them, and then await all of them?